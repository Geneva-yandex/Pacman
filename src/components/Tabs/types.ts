import {PropsWithChildren} from 'react';

export interface TabsProps extends PropsWithChildren<unknown> {
    selectedTab: string;
    onSelect: (tabId: string) => void;
}

export interface TabProps extends PropsWithChildren<unknown> {
    name: string;
    title: string;
    disabled?: boolean;
}
