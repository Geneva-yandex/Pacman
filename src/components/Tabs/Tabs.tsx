import * as React from 'react';
import {ReactElement} from 'react';
import bem from 'easy-bem';
import classnames from 'classnames';
import {TabProps, TabsProps} from './types';

const b = bem('Tabs');

export const Tab = (props: TabProps): ReactElement => {
    return props.children as ReactElement<any>;
};

const renderTabHeaderItem = ({name, title, disabled}: TabProps, selected: string, onClick: () => void): ReactElement => {
    const c = b('header-item');
    const className = classnames(c, {
        [`${c}--active`]: name === selected
    });

    return <li key={name} className={className}>
        <button className={b('tab-button')} onClick={onClick} disabled={disabled}>{title}</button>
    </li>;
};

export const Tabs = ({children, selectedTab, onSelect}: TabsProps): ReactElement => {
    const content = React.Children.map(children, (tab: ReactElement<TabProps>) => {
        return selectedTab === tab.props.name ? tab : null;
    });

    return <div className={b()}>
        <ul className={b('nav')}>
            {React.Children.map(children, (tab: ReactElement<TabProps>, index: number) => {
                return <li key={index} className={classnames({active: selectedTab === tab.props.name})}>
                    {renderTabHeaderItem(tab.props, selectedTab, () => onSelect(tab.props.name))}
                </li>;
            })}
        </ul>

        <div className={b('content')}>
            {content}
        </div>
    </div>;
};
