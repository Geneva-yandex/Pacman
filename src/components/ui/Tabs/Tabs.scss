@import "src/styles/variables";

.Tabs {
  &__nav {
    display: inline-flex;
    margin-bottom: 32px;
    border-bottom: 1px solid rgba(#eaeef4, 30%);
  }

  &__header-item {
    color: rgba(var(--color-text), 0.5);
    border-bottom: 3px solid transparent;
    margin-bottom: -1px;
    transition: 0.25s ease;

    &--active {
      color: var(--color-primary);
      border-color: var(--color-primary);
    }
  }

  &__tab-button {
    color: currentColor;
    padding: 16px 20px;
    font-size: 16px;
    font-weight: normal;
    transition: 0.25s ease;

    &:hover {
      color: var(--color-primary);
    }
  }
}
