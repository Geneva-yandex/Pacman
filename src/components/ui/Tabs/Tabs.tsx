import React from 'react';
import {ReactElement} from 'react';
import bem from 'easy-bem';
import {TabProps, TabsProps} from './types';
import './Tabs.scss';

const b = bem('Tabs');

export const Tab = (props: TabProps): ReactElement => {
    return props.children as ReactElement<any>;
};

const renderTabHeaderItem = ({name, title, disabled}: TabProps, selected: string, onClick: () => void): ReactElement => {
    const className = b('header-item', {active: name === selected});

    return <li key={name} className={className}>
        <button className={b('tab-button')} onClick={onClick} disabled={disabled}>{title}</button>
    </li>;
};

export const Tabs = ({children, selectedTab, onSelect}: TabsProps): ReactElement => {
    const headers: ReactElement[] = [];

    const content = React.Children.map(children, (tab: ReactElement<TabProps>) => {
        const tabHeader = renderTabHeaderItem(tab.props, selectedTab, () => onSelect(tab.props.name));
        headers.push(tabHeader);

        return selectedTab === tab.props.name ? tab : null;
    });

    return <div className={b()}>
        <ul className={b('nav')}>
            {headers.map(header => header)}
        </ul>

        <div className={b('content')}>
            {content}
        </div>
    </div>;
};

