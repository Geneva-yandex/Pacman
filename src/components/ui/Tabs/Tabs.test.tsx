import * as React from 'react';
import renderer from 'react-test-renderer';
import {Tabs, Tab} from './Tabs';

describe('Tabs', () => {
    test('Tabs component renders correctly', () => {
        const renderedComponent = renderer
            .create(
                <Tabs selectedTab={'firstTab'} onSelect={() => {}}>
                    <Tab name={'firstTab'} title='First tab'>
                        First tab content
                    </Tab>
                    <Tab name={'secondTab'} title='Second tab'>
                        Second tab content
                    </Tab>
                </Tabs>
            )
            .toJSON();

        expect(renderedComponent).toMatchSnapshot();
    });
});
