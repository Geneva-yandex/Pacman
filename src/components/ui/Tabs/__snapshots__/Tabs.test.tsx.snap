// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Tabs component renders correctly 1`] = `
<div
  className="Tabs"
>
  <ul
    className="Tabs__nav"
  >
    <li
      className="Tabs__header-item Tabs__header-item--active"
    >
      <button
        className="Tabs__tab-button"
        onClick={[Function]}
      >
        First tab
      </button>
    </li>
    <li
      className="Tabs__header-item"
    >
      <button
        className="Tabs__tab-button"
        onClick={[Function]}
      >
        Second tab
      </button>
    </li>
  </ul>
  <div
    className="Tabs__content"
  >
    First tab content
  </div>
</div>
`;
